js/ui/components/paygManager.js

```diff
@@ -0,0 +1,154 @@
+</interface> \
+</node>';
+
+var PaygManager = new Lang.Class({
+    Name: 'PaygManager',
+
+    _init: function() {
+        this._proxy = null;
+        this._proxyInfo = Gio.DBusInterfaceInfo.new_for_xml(EOS_PAYG_IFACE);
+
+        this._paygEnabled = false;
+        this._expiryTime = 0;
+
+        this._codeExpiredId = 0;
+        this._propertiesChangedId = 0;
+    },
+
+    _onPropertiesChanged: function(proxy, changedProps, invalidatedProps) {
+        let propsDict = changedProps.deep_unpack();
+        if (propsDict.hasOwnProperty('Enabled'))
+            this._setEnabled(this._proxy.Enabled);
+
+        if (propsDict.hasOwnProperty('ExpiryTime'))
+            this._setExpiryTime(this._proxy.ExpiryTime);
+    },
+
+    _setEnabled: function(value) {
+        if (this._paygEnabled === value)
+            return;
+
+        this._paygEnabled = value;
+        this.emit('payg-enabled', this._paygEnabled);
+    },
+
+    _setExpiryTime: function(value) {
+        if (this._expiryTime === value)
+            return;
+
+        this._expiryTime = value;
+        this.emit('payg-expiry-time-changed', this._expiryTime);
+    },
+
+    _onCodeExpired: function(proxy) {
+        this.emit('payg-code-expired');
+    },
+
+    _onProxyConstructed: function(object, res) {
+        try {
+            object.init_finish (res);
+        } catch (e) {
+            logError(e, "Error while constructing DBus proxy for " + EOS_PAYG_NAME);
+            return;
+        }
+
+        this._paygEnabled = this._proxy.Enabled;
+        this._expiryTime = this._proxy.ExpiryTime;
+    },
+
+    enable: function() {
+        if (!this._proxy) {
+            this._proxy = new Gio.DBusProxy({ g_connection: Gio.DBus.system,
+                                              g_interface_name: this._proxyInfo.name,
+                                              g_interface_info: this._proxyInfo,
+                                              g_name: EOS_PAYG_NAME,
+                                              g_object_path: EOS_PAYG_PATH,
+                                              g_flags: Gio.DBusProxyFlags.DO_NOT_AUTO_START_AT_CONSTRUCTION })
+
+            this._proxy.init_async(GLib.PRIORITY_DEFAULT, null, this._onProxyConstructed.bind(this));
+        }
+
+        this._propertiesChangedId = this._proxy.connect('g-properties-changed', this._onPropertiesChanged.bind(this));
+        this._codeExpiredId = this._proxy.connectSignal('Expired', this._onCodeExpired.bind(this));
+
+        Main.paygManager = this;
+    },
+
+    disable: function() {
+        Main.paygManager = null;
+
+        if (this._propertiesChangedId > 0) {
+            this._proxy.disconnect(this._propertiesChangedId);
+            this._propertiesChangedId = 0;
+        }
+
+        if (this._codeExpiredId > 0) {
+            this._proxy.disconnectSignal(this._codeExpiredId);
+            this._codeExpiredId = 0;
+        }
+
+        this._paygEnabled = false;
+        this._expiryTime = 0;
```

Comment:
Should these property value changes cause signal emissions (of `enabled` and `expiry-time-changed`)?

Reply:
Hmmm... I suppose so, although it's a bit weird to emit that signal when you're disabling the component, no? In which case, maybe I should not re-set the values at all, and simply let `_onProxyConstructed()` to set them again when re-enabling the component (emitting the signal there if the value for any of those values changed compared to when it was previously enabled).

Reply:
`disable()` is just another method call; it’s not like this is the `dispose` handler for the object and the object will no longer exist afterwards. Just in case someone calls `disable()` and then various UI/property bindings continue to use `paygEnabled` and `expiryTime`, I think it makes sense to reset the values (as you have done) and to emit the change signals (because all changes to the values should be notified). It makes sense to emit the change signals from `_onProxyConstructed()` too, for the same reasons.

Reply:
Fair point. I think that makes sense, I'll change it back.